src/stream/src/executor/hash_agg.rs

```diff
@@ -274,53 +286,69 @@ impl<K: HashKey, S: StateStore> HashAggExecutor<K, S> {
-                        Ok::<_, StreamExecutorError>((key.clone(), agg_group))
-                    })
+                vars.stats.total_lookup_count += 1;
+                if vars.dirty_groups.contains_key(key) {
+                    // already dirty
+                    return None;
+                }
+                match vars.agg_group_cache.get_mut(key) {
+                    Some(mut agg_group) => {
+                        let agg_group: &mut Option<_> = &mut agg_group;
+                        assert!(
+                            agg_group.is_some(),
+                            "invalid state: AggGroup is None in cache but not dirty"
+                        );
+                        // move from cache to `dirty_groups`
+                        vars.dirty_groups
+                            .insert(key.clone(), agg_group.take().unwrap());
+                        None // no need to create
+                    }
+                    None => {
+                        vars.stats.lookup_miss_count += 1;
+                        Some(async {
+                            // Create `AggGroup` for the current group if not exists. This will
+                            // restore agg states from the intermediate state table.
+                            let agg_group = AggGroup::create(
+                                Some(GroupKey::new(
+                                    key.deserialize(group_key_types)?,
+                                    Some(this.group_key_table_pk_projection.clone()),
+                                )),
+                                &this.agg_calls,
+                                &this.agg_funcs,
+                                &this.storages,
+                                &this.intermediate_state_table,
+                                &this.input_pk_indices,
+                                this.row_count_index,
+                                this.extreme_cache_size,
+                                &this.input_schema,
+                            )
+                            .await?;
+                            Ok::<_, StreamExecutorError>((key.clone(), agg_group))
+                        })
+                    }
                 }
             })
             .collect_vec(); // collect is necessary to avoid lifetime issue of `agg_group_cache`
 
-        stats.chunk_total_lookup_count += 1;
+        vars.stats.chunk_total_lookup_count += 1;
         if !futs.is_empty() {
             // If not all the required states/keys are in the cache, it's a chunk-level cache miss.
-            stats.chunk_lookup_miss_count += 1;
+            vars.stats.chunk_lookup_miss_count += 1;
             let mut buffered = stream::iter(futs).buffer_unordered(10).fuse();
             while let Some(result) = buffered.next().await {
                 let (key, agg_group) = result?;
-                cache.put(key, agg_group);
+                vars.dirty_groups.insert(key, Box::new(agg_group));
```

Comment:
What about using `try_insert` to assert that the key does not exist?

Reply:
A little bit doubt on the necessity, the code looks simple enough here. Changing here will also suggest changing the `insert` call in the first match arm, and that will make it quite verbose. I added an assertion on the ret value of `insert`.